### PR TITLE
Theme detection and update improvements

### PR DESCRIPTION
**Pull request by Chris-N-K**
Hi, 
as mentioned in the issue #236 I did a first draft of a possible implementation of dynamic styling based on the `napari` viewer theme.

This rework completely omits the concept of static style sheets for the different `napari` theme and instead translates the current viewer theme into a `matplotlib` compatible style dictionary.

This happens upon widget initiation and the style dictionary is stored in the attribute: `napari_theme_style_sheet`.
The style dictionary can be directly used for a `motplotlib.style.contect`.
Additionally, I fixed the viewer theme callback and the style change upon theme change works now properly.

I encountered some strange during the rework, if I set the `facecolor` of the figure or axes to something other than transparent (none) the color will stay. Seems like the `facecolor`s are not updated by redraws. I'm feel this is a `matplotlib` bug, but I could be wrong, maybe I just used the wrong functions for redrawing the figure and axis faces.
EDIT:
I can reproduce this behaviour with pure `matplotlib`, if I explicitly set the facecolor and redraw it changes, but if I use a style context it stays the same.

If you find the implementation suitable you or I would need to give the test an update as well.

### CODE REVIEW

**Comment by dstansby** [comment]
pre-commit.ci autofix

**Comment by dstansby** [comment]
I really like this approach! Not having to deal with custom matplotlib style sheet paths makes life a lot simpler for us, and also gives the user only one place to customise the theme instead of having to do it for both napari and napari-matplotlib. I need to do some manual testing, but in the meantime would you be happy chaging either the code or tests to make the tests pass?

I automatically fixed formatting issues, so there's now an extra commit on your branch. If you want to do taht yourself locally you can install and run pre-commit: https://pre-commit.com/

**Comment by Chris-N-K** [comment]
Hi, I'm happy you like my suggestions. I can take a look at the tests but I'm rather short on time at the moment so this could take some time.
If you are fine with taking a look at the image tests, I would go through the rest.

**Comment by dstansby** [comment]
I've updated the image tests, and cleaned out the old styling code that isn't needed any more. I still need to debug some of the new images, as the text colour doesn't look right. Manual testing indicates this is looking good otherwise though 👍

File: docs/user_guide.rst [hunk 1]
```diff
@@ -40,11 +40,6 @@ To use these:
...
 Customising plots
 -----------------
-`Matplotlib style sheets <https://matplotlib.org/stable/tutorials/introductory/customizing.html#defining-your-own-style>`__ can be used to customise
-the plots generated by ``napari-matplotlib``.
-To use a custom style sheet:
-
-1. Save it as ``napari-matplotlib.mplstyle``
-2. Put it in the Matplotlib configuration directory.
-   The location of this directory varies on different computers,
-   and can be found by calling :func:`matplotlib.get_configdir()`.
```
Comment (samcunliffe) [inline docs/user_guide.rst:50]: 👍 
Does this warrant a major version bump?

File: src/napari_matplotlib/base.py [hunk 2]
```diff
@@ -45,18 +42,17 @@ def __init__(
...
-        with mplstyle.context(self.mpl_style_sheet_path):
+        with mplstyle.context(self.napari_theme_style_sheet):
             self.canvas = FigureCanvasQTAgg()  # type: ignore[no-untyped-call]
 
         self.canvas.figure.set_layout_engine("constrained")
         self.toolbar = NapariNavigationToolbar(self.canvas, parent=self)
         self._replace_toolbar_icons()
-        # callback to update when napari theme changed
-        # TODO: this isn't working completely (see issue #140)
-        # most of our styling respects the theme change but not all
```
Comment (samcunliffe) [inline src/napari_matplotlib/base.py:59]: Fixes #140 ?

File: docs/user_guide.rst (hunk 1 above)
Comment (dstansby) [reply docs/user_guide.rst:50]: Yes, hence the 2.0 milestone 😉

File: src/napari_matplotlib/base.py (hunk 2 above)
Comment (Chris-N-K) [reply src/napari_matplotlib/base.py:59]: Yes it fixes the axelabel text color issue, but is kind of a work around for the background issue.
The background is transparent, which is not optimal in all cases, but for what ever reason I was not able to implement it in a way that the background is updated correctly.
I guess there might be an underlying bug not directly connected to napari-matplotlib but either napari or and this is my guess the way matplotlib updates the canvas.

**Comment by Chris-N-K** [comment]
Looks like all the tests work now. Is there still something I should check?

**Comment by dstansby** [comment]
pre-commit-ci autofix

**Comment by dstansby** [comment]
Nope, I think this all looks good now 👍 . I'll wait for the tests to pass again then merge. Will try and get this in a new release in the next few weeks or so, thanks a lot for the PR!

**Comment by Chris-N-K** [comment]
Happy I could help ^^

**Pull request merged**